Show marked/edited markers, without swipe gesture

PIG208 opened the pull request:
<del>This is still a WIP.</del>

The swipe gesture is implemented such that the marker is
- expanded when there is horizontal movement to the right
- stretched following the touch movement
- retracted when the touch is released

Notably, the content box still re-wraps as the marker pushes it to the right when it expands.To fix this, the content box might need to have a fixed size so the text can go off-screen and remain unchanged. This turned out to be tricky to accomplish.

This is stacked on top of #750. It fixes #616 and fixes part of #171.

https://github.com/zulip/zulip-flutter/assets/39874143/1d57d454-e393-4e2a-a9b8-170d6d328c35

gnprice commented:
Thanks! Looking forward to having this.

@terpimost @alya for feedback on how the swipe gesture behaves from a user perspective.

> Notably, the content box still re-wraps as the marker pushes it to the right when it expands.To fix this, the content box might need to have a fixed size so the text can go off-screen and remain unchanged. This turned out to be tricky to accomplish.

For this I think the key is, instead of a `Row`, to use something like a [`SlideTransition`](https://main-api.flutter.dev/flutter/widgets/SlideTransition-class.html) (which is what `Dismissible` uses). If the coordinate system that one uses feels inconvenient, then [`PositionedTransition`](https://main-api.flutter.dev/flutter/widgets/PositionedTransition-class.html) might be cleaner. (And we can discuss in more detail in a chat thread.)

alya commented:
Pushing the context so that part of it gets hidden, rather than rewrapping, makes sense to me.

PIG208 commented:
The text re-wrapping issue has been fixed. I have also added a responsive fix to the star alignment issue. It's a draft because I still need to add tests for it.

gnprice commented:
Great.

I was using the app this afternoon and happened to have a build from this PR on my phone, and found I really appreciated having the new information about which messages had been edited.

That caused me to think of an idea for how we can split this into a part that delivers the key functionality and that we can merge soon, and a more complex part that can come after it as a follow-up:
* For a first version, we could show just the icons, without the swipe-to-reveal feature.

  That cuts out the gesture handling, the animation, and all the tricky bits about coordinating how the marker's layout and colors change in synchrony with the motion.

  The part that remains will be exactly how we intend the final feature to look in steady state, whenever the user isn't actively pulling on one of these messages to reveal the marker. So it'll also be a good way for us to focus on polishing that design to exactly how we want it, before introducing the complexity of the swipe animation.

  From a user perspective, the one thing that's missing in this version is that it may not be super discoverable what the icons mean. But once you figure it out (perhaps by seeing them on a thread where you have other clues), you know; there's only two of them to learn.

* Then a subsequent PR introduces swipe-to-reveal and the colored markers.

I see the UI logic in this revision looks a lot cleaner, so that follow-up might happen soon after. Even then, though, I think the split will be helpful for reducing the per-PR complexity for review. So let's try making that split.

PIG208 commented:
I extracted a diff that implements the animation to #765. This PR should be ready for review now.

PIG208 commented:
Rebased and added some test cases for the UI change.

chrisbobbe commented:
GitHub is telling me that some conflicts have appeared; would you resolve those please?

PIG208 commented:
Just rebased the PR.

chrisbobbe commented on lib/widgets/edit_state_marker.dart:
```diff
@@ -25,11 +25,11 @@ class _SwipableMessageRowState extends State<SwipableMessageRow> {
+      padding: const EdgeInsets.only(top: 5.5),
+      child: Row(children: [
+          const Text("", style: TextStyle(fontSize: 15)),
+          Icon(ZulipIcons.star_filled, size: 16, color: theme.starColor)]));
```
So normally, to align text-like elements by their baselines, we would use a `Row` with certain arguments passed:

```dart
Row(
  crossAxisAlignment: CrossAxisAlignment.baseline,
  textBaseline: localizedTextBaseline(context),
```

I don't know of a way to do that with a `Stack` instead of a `Row`, though. Do you think you could make it work—including with the animation—using a `Row`, where the first child is the edited/moved marker, the second is the message, and the third is the star? You may need to give the message area a fixed size, but I think that could be done with a `LayoutBuilder`, as in the diff below.

I haven't thoroughly tested this, and I don't know how it might interact with the animation requirement—but an advantage is that we'll keep the baseline alignment even as the message's first line varies in height. Screenshots below, showing a message starting with the line `# heading 1`, which makes a tall h1 element:

<details>
<summary>diff atop the PR branch</summary>

```diff
diff --git lib/widgets/edit_state_marker.dart lib/widgets/edit_state_marker.dart
index fdd40a43b..37fc93c7c 100644
--- lib/widgets/edit_state_marker.dart
+++ lib/widgets/edit_state_marker.dart
@@ -4,6 +4,7 @@ import 'package:flutter_gen/gen_l10n/zulip_localizations.dart';
 
 import '../api/model/model.dart';
 import 'icons.dart';
+import 'text.dart';
 import 'theme.dart';
 
 class SwipableMessageRow extends StatefulWidget {
@@ -25,26 +26,23 @@ class _SwipableMessageRowState extends State<SwipableMessageRow> {
   Widget build(BuildContext context) {
     final theme = Theme.of(context).extension<DesignVariables>()!;
 
-    var star = Padding(
-      padding: const EdgeInsets.only(top: 5.5),
-      child: Row(children: [
-          const Text("", style: TextStyle(fontSize: 15)),
-          Icon(ZulipIcons.star_filled, size: 16, color: theme.starColor)]));
     final hasMarker = widget.message.editState != MessageEditState.none;
 
-    return Stack(
-      children: [
-        if (hasMarker) Positioned(
 
-    var star = Padding(
-      padding: const EdgeInsets.only(top: 5.5),
-      child: Row(children: [
-          const Text("", style: TextStyle(fontSize: 15)),
-          Icon(ZulipIcons.star_filled, size: 16, color: theme.starColor)]));
     final hasMarker = widget.message.editState != MessageEditState.none;
 
-    return Stack(
-      children: [
-        if (hasMarker) Positioned(
-          left: 0,
-          child: EditStateMarker(editState: widget.message.editState)),
-        Padding(
-          padding: const EdgeInsets.only(left: 16, right: 16),
-          child: widget.child),
-        if (widget.message.flags.contains(MessageFlag.starred))
-          Positioned(
-            right: 0,
-            child: star),
-      ],
+    return LayoutBuilder(
+      builder: (context, constraints) => Row(
+        crossAxisAlignment: CrossAxisAlignment.baseline,
+        textBaseline: localizedTextBaseline(context),
+        children: [
+          hasMarker
+            ? EditStateMarker(editState: widget.message.editState)
+            : const SizedBox(width: 16),
+          SizedBox(width: constraints.maxWidth - 32, child: widget.child),
+          SizedBox(width: 16,
+            child: widget.message.flags.contains(MessageFlag.starred)
+              ? Icon(ZulipIcons.star_filled, size: 16, color: theme.starColor)
+              : null),
+        ],
+      ),
     );
   }
```
</details>

| This PR | with `Row` |
| --- | --- |
| ![E1B3BDED-0E58-425A-ABC3-7B646322F6BA](https://github.com/zulip/zulip-flutter/assets/22248748/0cb3a136-0425-48ad-9f42-a51b714f60d2) | ![CB5DDBAC-8715-44B2-92D0-1C26351110FE](https://github.com/zulip/zulip-flutter/assets/22248748/29460d13-6f7d-4677-ae7a-ec9934d15135) |

PIG208 replied:
https://github.com/zulip/zulip-flutter/assets/39874143/e5a74684-e444-4f80-bb70-6e48f44a3368

Wrapping the child with Transform.translate seems to work. We just need to deal with it overflowing the Row.

chrisbobbe commented on lib/widgets/edit_state_marker.dart:
```diff
@@ -0,0 +1,127 @@
+    //   https://www.figma.com/file/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=813%3A28817&mode=dev .
+    var star = Padding(padding: const EdgeInsets.only(top: 5.5),
+      child: Icon(ZulipIcons.star_filled, size: 16, color: theme.starColor));
+    final hasMarker = widget.message.editState != MessageEditState.none;
```
Hmm, this file is called edit_state_marker.dart, so I'm wondering if the message-star UI code quite belongs in it. Maybe we want a different name for the file? @gnprice might have thoughts on the organization here.

PIG208 replied:
Yeah "edit_state_marker" is not a great fit. We could name it as with related to `SwipableMessageRow`, which is the only widget supposed to be exposed.

gnprice replied:
How about having the caller stay responsible for the star? Then the child it passes here can include the star as well as the message content and reactions.

chrisbobbe commented on test/widgets/message_list_test.dart:
```diff
@@ -576,6 +576,39 @@ void main() {
+      await tester.pump();
+      checkMarkersCount(edited: 1, moved: 0);
+
+      await store.handleEvent(eg.updateMessageMoveEvent
+        ([message, message2], origTopic: 'old', newTopic: 'new'));
```
I don't think I've seen a method call split between its name and its opening `(`; I don't think we want that. :)

chrisbobbe reviewed (commented):
Thanks, this will be a useful feature! Comments below, including a curiosity about improving the star's (and the edit/move marker's) alignment with the message content.

PIG208 commented on lib/widgets/theme.dart:
```diff
@@ -81,7 +81,11 @@ class DesignVariables extends ThemeExtension<DesignVariables> {
     title = const Color(0xff1a1a1a),
-    streamColorSwatches = StreamColorSwatches.light;
+    streamColorSwatches = StreamColorSwatches.light,
+    starColor = const HSLColor.fromAHSL(0.5, 47, 1, 0.41).toColor(),
+    bgMarker = const Color(0xffddecf6),
+    textMarker = const Color(0xff26516e),
+    textMarkerLight = const Color(0xff92a7b6);
```
Not sure if edit permission is required for me to see the local variables. Sent an edit request.

PIG208 replied:
Pushed a revision without the theme changes for star and other colors used for the marker.

gnprice replied:
I think it's fine to use the same class to house variables that aren't named in that list in Figma.

Let's just keep it organized in a way that makes that clear. So in particular have such variables' definitions in their own stanza, with a comment. Like this:
```dart
  // Not exactly from the Figma design, but from Vlad anyway.
  final StreamColorSwatches streamColorSwatches;

  // Not named variables in Figma; taken from older Figma drafts, or elsewhere.
  final Color starColor;
  final Color bgMarker;
  final Color textMarker;
  final Color textMarkerLight;
```

gnprice replied:
> Not sure if edit permission is required for me to see the local variables. Sent an edit request.

I think I just approved that request. If you're still having trouble or have any questions, let's discuss in `#mobile-team` — I can try poking at other settings as an admin, and Chris might have other tips or Vlad might chime in with advice.

PIG208 replied:
Reverted the theme-less change. With the permission the local variables approach now works for me. Thanks!

PIG208 commented:
Thanks for the review! I pushed some updates addressing some reviews, but there are still some open discussions on matters like how we organize the theme variables and the edit marker code.

PIG208 commented:
![Screenshot_20240705_155215.jpg](https://github.com/zulip/zulip-flutter/assets/39874143/55cacc73-e454-4a3c-8e5f-e00df5aa90fb)

![Screenshot_20240705_155156.jpg](https://github.com/zulip/zulip-flutter/assets/39874143/7833e098-f8ae-48d6-944d-4dff714eed3b)

PIG208 commented:
I have refactored the star out of `SwipableMessageRow` and re-implemented the thing with the `Row`-based approach drafted by @chrisbobbe. It works fine without animation.

While adding animation, it turned out that the overflow issue was pretty tricky, but solvable by wrapping the entire widget with an `OverflowBox` within the `LayoutBuilder`, and constraining the size of the message content from the parent (see #765).

gnprice commented:
> While adding animation, it turned out that the overflow issue was pretty tricky, but solvable by wrapping the entire widget with an `OverflowBox` within the `LayoutBuilder`, and constraining the size of the message content from the parent (see #765).

Interesting. I hope we can avoid needing a LayoutBuilder; but we'll cross that bridge when we get to it.

For this PR I'm happy if we merge something that's clean and has the right static behavior, even if it seems like it makes a LayoutBuilder or other tricks necessary for #765. After this is done and we focus on #765, maybe we'll find a way to avoid LayoutBuilder within the same architecture we merged in this PR; or maybe we'll do so but have to rework some of that architecture; or maybe we'll decide we need the LayoutBuilder after all.

chrisbobbe commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,115 @@
+        decoration: BoxDecoration(
+          borderRadius: BorderRadius.circular(2),
+          color: theme.bgMarker.withAlpha(0)),
+        child: Padding(
+          padding: const EdgeInsets.all(1.0),
+          child: marker),
```
I don't see vertical padding in the Figma. Am I missing it?

The height of the rounded-rectangle gets set by the line of text (e.g. "Edited"), it looks like. That's 22, because the line-height is set to that, and I don't see anything creating more vertical whitespace, in or out of the rounded-rectangle.

I do see some horizontal padding, but I wonder if this is the right way to express it. Here's a screenshot of the Figma:

<img width="983" alt="image" src="https://github.com/zulip/zulip-flutter/assets/22248748/d75026d9-e69c-4431-88a0-1d155f34d019">

In collapsed state, I see a 3px horizontal gap between the icon and the message content.

In uncollapsed state, I see:

a) 4px between the rounded-rectangle and the message content
b) 2px between the icon and the right edge of the rounded-rectangle
c) 5px between "Edited" and the icon
d) 3px between the left edge of the rounded-rectangle and "Edited"

To my eye, (a) should maybe just become 3px, to match the 3px in the collapsed state. (Maybe helps keep the implementation simple too?) And it seems to me like (b) and (d) should equal each other, to be symmetrical; not sure which value is better. Then since the uncollapse action adds space between the icon and the message content, perhaps (b) should animate from 0 to 2 as part of the animation (and then we animate (d), too, to be symmetrical).

Also, let's do (c) explicitly with a 5px gap, instead of a space character as in this revision. That might be something to animate from zero too, I'm not sure.

PIG208 replied:
Thanks for the breakdown on the design. Really helpful! I'm moving the relevant updates for this to #765.

chrisbobbe reviewed (commented):
Thanks! Comments below.

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -962,25 +963,22 @@ class MessageWithPossibleSender extends StatelessWidget {
+                  ? Icon(ZulipIcons.star_filled, size: 16, color: theme.starColor)
+                  : null),
+            ],
+          ),
```
Yeah, separating those would definitely be helpful for review. They seem like they should be naturally pretty independent changes, too.

chrisbobbe commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,115 @@
+        OverflowBox(
+          fit: OverflowBoxFit.deferToChild,
+          maxWidth: 10,
+          child: Icon(icon, size: iconSize, color: theme.textMarkerLight),
```
(Also, I don't think we have a source for the edited-moved marker colors in dark theme yet. So let's put a `// TODO(#95) need dark-theme color` on them.)

PIG208 replied:
Yeah, I think with 50% it matches the design better.
<details>

<summary>Color comparsion</summary>

| Before | After (50% opacity) |
| --- | --- |
| ![93b790eb-912a-47d7-a2c1-1d2970356c5e](https://github.com/zulip/zulip-flutter/assets/39874143/e7a9a535-e5b1-46a9-bd00-d209f0c21d22) | ![6e6c3d4d-9395-4201-b967-191ffa736d4c](https://github.com/zulip/zulip-flutter/assets/39874143/d81eb490-9ff8-4ea7-bef0-6ae3f0f6d952)

</details>

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,115 @@
+      children: [
+        Flexible(
+          fit: FlexFit.loose,
+          // For now, [markerText] is not displayed because it is transparent and
+          // there is not enough space in the parent ConstrainedBox.
+          child: Text('$markerText ',
```
Let's leave this out of this static version, since it's not yet being used — that will make the logic easier to think about. Then we can introduce it (modulo other edits, like Chris's comment about the space character vs. an explicit number of pixels of spacing) in the followup PR #765 that adds the swipe-to-reveal feature.

PIG208 commented:
Thanks for the reviews! This has been updated and is ready for another round now.

chrisbobbe commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,86 @@
+    return ConstrainedBox(
+      constraints: const BoxConstraints(maxWidth: widthCollapsed),
+      child: Padding(
+        padding: const EdgeInsets.only(left: 5, right: 3),
```
Since left and right are unequal, this is a place to use `EdgeInsetsDirectional` instead of `EdgeInsets`.

chrisbobbe commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,86 @@
+}
+
+class _EditStateMarker extends StatelessWidget {
+  /// The minimum width of the marker.
+  //
+  // Currently, only the collapsed state of the marker has been implemented,
+  // where only the marker icon, not the marker text, is visible.
```
nit: empty line unnecessary

```suggestion
  /// The minimum width of the marker.
  // Currently, only the collapsed state of the marker has been implemented,
  // where only the marker icon, not the marker text, is visible.
```

chrisbobbe commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,86 @@
+      constraints: const BoxConstraints(maxWidth: widthCollapsed),
+      child: Padding(
+        padding: const EdgeInsets.only(left: 5, right: 3),
+        child: OverflowBox(
+          fit: OverflowBoxFit.deferToChild,
+          maxWidth: 8,
+          child: Icon(icon, size: iconSize,
+            color: designVariables.editedMovedMarkerCollapsed)),
+        ),
+    );
```
nit:

```suggestion
        child: OverflowBox(
          fit: OverflowBoxFit.deferToChild,
          maxWidth: 8,
          child: Icon(icon, size: iconSize,
            color: designVariables.editedMovedMarkerCollapsed))));
```

chrisbobbe reviewed (commented):
Thanks! LGTM except a few small comments below, and I'll mark this as ready for Greg's review.

chrisbobbe commented:
I see you've pushed changes; is this ready for Greg's review? I took a look at what you pushed, and it resolves my previous review. Thanks!

PIG208 commented:
Yes, this is ready for review.

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,83 @@
+        break;
+      case MessageEditState.moved:
+        icon = ZulipIcons.message_moved;
+        iconSize = 8;
```
It looks like the underlying icon SVG has a height of 8 but a width of 9. Will this cause it to be scaled down to fit into an 8x8 box?

PIG208 replied:
Yeah, that should be the case.

| | moved (8x8) | edited (8x14) |
| - | - | - |
| Flutter | ![Screenshot from 2024-07-12 21-02-36](https://github.com/user-attachments/assets/50400d8c-1e20-470c-bfa1-c03c57739e6a) | ![Screenshot from 2024-07-12 21-01-17](https://github.com/user-attachments/assets/35b332ec-77b0-460a-9f5a-7073df90cbc0) |
| Figma | ![Screenshot from 2024-07-12 21-07-57](https://github.com/user-attachments/assets/9a2bb83a-46ba-4bb9-a97f-093230b8c753) | ![Screenshot from 2024-07-12 21-08-10](https://github.com/user-attachments/assets/01f9a807-6190-4a3c-8ef9-f95e9ff8af1d) |

`iconSize` scales both the width and height of the icon, until it exceeds the constraints of the parent.

<details>

![image](https://github.com/user-attachments/assets/30924df4-97fb-45b6-991f-e238e56aa6e9)

</details>

gnprice replied:
Hmm I see - so the icon file has a width of 9 but the design gives it a width of 8. Confusing. Worth clarifying in a comment, at the least.

Probably also good to ask Vlad about it, in a chat thread in... #mobile, I guess, since it's design (even though it's a fine detail of design).

PIG208 replied:
I checked again. So it looks like a later edition has an icon of size 8x8 when downloaded. But I'm starting to think if we should use the frame (8x14) containing that icon, not just the icon itself (8x8) for the design. That way the size also matches the "edited" marker icon.

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,83 @@
+    return ConstrainedBox(
+      constraints: const BoxConstraints(maxWidth: widthCollapsed),
+      child: Padding(
+        padding: const EdgeInsetsDirectional.only(start: 5, end: 3),
+        child: OverflowBox(
+          fit: OverflowBoxFit.deferToChild,
+          maxWidth: 8,
```
Can we do this without OverflowBox?

There's no real overflow happening in the end, right? The wider icon, for moved, is only width 9 — so it should fit just fine in the 16px width that this Padding widget has. Can we adjust the padding to match that instead?

PIG208 replied:
That's doable. I think `OverflowBo`x is something that we might need for the animated version.

gnprice replied:
OK. Even if we end up needing it for animated, I'd like to try starting without it as a baseline - I think that'll make this easier to reason about.

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -976,13 +975,9 @@ class MessageWithPossibleSender extends StatelessWidget {
                 ])),
             SizedBox(width: 16,
```
nit: should follow indentation of other child

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -964,10 +963,10 @@ class MessageWithPossibleSender extends StatelessWidget {
           if (senderRow != null)
             Padding(padding: const EdgeInsets.fromLTRB(16, 2, 16, 0),
               child: senderRow),
-          Row(crossAxisAlignment: CrossAxisAlignment.start, children: [
```
msglist: Fix star icon alignment issue.
    
    By aligning the row using `textBaseline`, we no longer need to apply
    the padding to the star.

Neat!

gnprice reviewed (commented):
Thanks @PIG208, and thanks @chrisbobbe for the previous reviews! Looking forward to having this feature.

I'm only partway through the last commit and it's time for me to go home, so I'll leave the rest of that commit for a later review. Comments below for so far.

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -964,25 +963,20 @@ class MessageWithPossibleSender extends StatelessWidget {
+              SizedBox(width: 16,
+                child: message.flags.contains(MessageFlag.starred)
+                  ? Icon(ZulipIcons.star_filled, size: 16, color: designVariables.star)
+                  : null)]),
```
nit:
```suggestion
                  : null),
            ]),
```

(The pattern is that when there's a list, so the items are peers or siblings of each other, the last item gets a trailing comma just like the other items and the closing delimiter like `]` goes on the next line.

When there's one item that's the main payload and the other items are configuration around that — so in particular, for the `child` parameter vs. the other parameters of some widget constructor — then the main payload goes at the end and piles up the closing `)` delimiters without spilling onto more lines.)

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+    required this.message,
+  });
+
+  final List<Widget> children;
+  final Message message;
```
nit: "child" or "children" fields on widgets come after other fields (here and in constructor parameters… and in constructor calls, as mentioned on another comment just now)

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+  final List<Widget> children;
+  final Message message;
+
+  @override
+  State<StatefulWidget> createState() => _SwipableMessageRowState();
+}
+
+class _SwipableMessageRowState extends State<SwipableMessageRow> {
+  @override
+  Widget build(BuildContext context) {
```
Seems like this can just as well be a stateless widget:
```suggestion
  @override
  Widget build(BuildContext context) {
```

(We'll need state here for animation, in #765, but it's easy to add then. And the state might go on a child, rather than this widget itself — one way or another I'd like to arrange it so we don't need a state in the case of MessageEditState.none, at least once the state starts holding other resources like an AnimationController.)

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+  /// The minimum width of the marker.
+  // Currently, only the collapsed state of the marker has been implemented,
+  // where only the marker icon, not the marker text, is visible.
+  static const double widthCollapsed = 16;
```
nit: this goes after the constructor and field:
https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#order-other-class-members-in-a-way-that-makes-sense

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+  static const double widthCollapsed = 16;
+
+  const _EditStateMarker({
+    required MessageEditState editState,
+  }) : _editState = editState;
```
nit: can simplify this by making the field public

The class is private anyway, so the visibility of the field doesn't matter.

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+  Widget build(BuildContext context) {
+    final designVariables = DesignVariables.of(context);
+
+    final IconData icon;
+    final double iconSize;
+
+    switch (_editState) {
```
nit: these two locals are tightly related to this `switch` (it's what sets them), so keep them adjacent in the same stanza:
```suggestion
    final IconData icon;
    final double iconSize;
    switch (_editState) {
```

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+      case MessageEditState.edited:
+        icon = ZulipIcons.edited;
+        iconSize = 14;
+        break;
+      case MessageEditState.moved:
```
nit:
```suggestion
      case MessageEditState.moved:
```

Dart has removed the need for `break` as a routine part of a switch statement, which is a nice language improvement. See [spec discussion](https://github.com/dart-lang/language/blob/2b3d78182c226aa26d46df341fc2cdb48a88cbaf/accepted/3.0/patterns/feature-specification.md#implicit-break).

It looks like there is a lint rule for this. Likely we should enable it:
https://dart.dev/tools/linter-rules/unnecessary_breaks

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+      children: [
+        hasMarker
+          ? _EditStateMarker(editState: widget.message.editState)
+          : const SizedBox(width: 16),
```
```suggestion
          : const SizedBox(width: _EditStateMarker.widthCollapsed),
```

It's important that this 16 is the same as the other 16, right? So let's make that explicit by referring to it by a common name.

gnprice commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+    final double iconSize;
+
+    switch (_editState) {
+      case MessageEditState.none:
+        return const SizedBox(width: widthCollapsed);
```
This case isn't intended to be possible, right? Let's make that explicit with an assert:
```suggestion
      case MessageEditState.none:
        assert(false);
        return const SizedBox(width: widthCollapsed);
```

gnprice commented on lib/widgets/theme.dart:
```diff
@@ -111,6 +123,12 @@ class DesignVariables extends ThemeExtension<DesignVariables> {
 
+  // Not named variables in Figma; taken from older Figma drafts, or elsewhere.
+  final Color star;
+  final Color editedMovedMarkerBg;
+  final Color editedMovedMarkerExpanded;
```
These two aren't used yet in this commit (or PR), right? Let's save them for the commit where they do get used:
```suggestion
```

gnprice commented on test/widgets/message_list_test.dart:
```diff
@@ -576,6 +576,39 @@ void main() {
+      checkMarkersCount(edited: 0, moved: 0);
+    });
+
+    testWidgets('edited and moved messages from events', (tester) async {
```
Let's have another test case that checks that the markers don't affect the layout of anything else. That seems like a class of bug that could easily be introduced in some future change to the message UI.

So that can look like:
* put three messages on the screen; the one we'll change the state of, and one each before and after it
* note the Rect position of everything relevant:
  * each of the three messages
  * the target message's star (make it starred so this appears)
  * some text in the target message
  * maybe other things I'm not thinking of
  * … oh a reaction, that's a good one (and there might still be more I'm not thinking of)
* then make the message be moved; and look at all the same locations and confirm none have changed
* then make the message be edited; and ditto

PIG208 replied:
Added the "edit state updates do not affect layout" test case. It uses a heavy helper function to record the items you mentioned, as well as `ImageNode`s and `ParagraphNode`s from the target message.

gnprice reviewed (commented):
OK, and here's the rest of a review (based on your latest revision, so including a couple of comments new with this revision).

Most are small. Among the more interesting: the first comment is one where I'm uncertain whether this gives the right layout. The last comment asks for a layout-related test.

PIG208 commented on lib/widgets/swipable_message_row.dart:
```diff
@@ -0,0 +1,80 @@
+    return ConstrainedBox(
+      constraints: const BoxConstraints(maxWidth: widthCollapsed),
+      child: Padding(
+        padding: const EdgeInsetsDirectional.only(start: 5, end: 3),
+        child: Icon(icon, size: iconSize,
```
I'm switching to use the 8x14 Figma frame that contains the move icon, with `EdgeInsetsDirectional.only(start: 2)` for the padding. The look appears to be the same when we were using `EdgeInsetsDirectional.only(start: 5, end: 3)`, but that clips the icon from 14 in width to 8 in width.

PIG208 commented:
Thanks for the review @gnprice ! The PR has been updated.

gnprice reviewed (commented):
Thanks @PIG208 for the revision! Comments below.

Would you also post an updated screenshot of how this looks with each of the two kinds of markers? I'm not sure if the images above are the same as in the current revision.

In particular when I look at an edited message, the bottom of the icon is aligned with the baseline of the (first line of the) message text… but then that leaves the top of the icon sticking up well above the ascender line of the text. The result is that it feels like the icon is sticking up above the message, so that it's not visually clear if it's meant to be attached to the message or what.

PIG208 commented on test/widgets/message_list_test.dart:
```diff
@@ -576,6 +577,107 @@ void main() {
+      check(find.byIcon(ZulipIcons.message_moved).evaluate()).length.equals(moved);
+    }
+
+    testWidgets('no edited or moved messages', (WidgetTester tester) async {
```
I see. I added those in the previous revision because I saw we doing that in the surrounding test cases.

gnprice replied:
Yeah, matching the surrounding code is a good instinct.

Like I said, having the explicit type here is pretty harmless. But maybe I'll do a quick sweep to remove them for `testWidgets` calls throughout our tests, just so it's uniformly a bit simpler. (From a quick scan, about 1/3 of our `testWidgets` calls have them.)

PIG208 commented on test/widgets/message_list_test.dart:
```diff
@@ -576,6 +577,107 @@ void main() {
+      final messages = [
+        eg.streamMessage(),
+        eg.streamMessage(
+          content: ContentExample.imageClusterThenContent.html,
```
I wanted to test that because I discovered the part with `MessageImageList` do not respond to swipe on the animated branch. It is not relevant now but might be later.

PIG208 replied:
This is something I should note in the commit message, apparently.

gnprice replied:
Interesting, that makes sense.

I think that still won't be relevant for this layout test, though — having the image interfere with a gesture (because the image has its own gesture recognizers) is a lot more natural than having it interfere with an ancestor's layout. So it's still probably best left out of this test, but then you might use this example in a new test in the swipe-to-reveal PR.

gnprice replied:
(Or even in that later test, it sounds like `imageSingle` would likely be enough to make the point — it has an image and even a paragraph too, just one of each instead of two of each.)

gnprice commented on test/widgets/message_list_test.dart:
```diff
@@ -576,6 +577,107 @@ void main() {
+            (i) => tester.getRect(finder.at(i)));
+        }
+
+        result.add({
+          'MessageWithPossibleSender': [tester.getRect(baseFinder)],
```
Oh, and I forgot to post what the error output looks like after that change. Here it is:
```
$ flutter test test/widgets/message_list_test.dart --name EditState
00:03 +2: EditStateMarker edit state updates do not affect layout                                        
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: a List<List<(String, Rect)>> that:
  is deeply equal to [[(MessageWithPossibleSender, Rect.fromLTRB(20.0, 187.0, 780.0, 259.0)),
  (MessageContent, Rect.fromLTRB(36.0, 225.0, 764.0, 255.0))],
  [(MessageWithPossibleSender, Rect.fromLTRB(20.0, 259.0, 780.0, 457.0)),
  (MessageContent, Rect.fromLTRB(36.0, 263.0, 764.0, 430.0)),
  (Star, Rect.fromLTRB(764.0, 270.3, 780.0, 286.3)),
  (ReactionChipsList, Rect.fromLTRB(36.0, 430.0, 764.0, 453.0))],
  [(MessageWithPossibleSender, Rect.fromLTRB(20.0, 457.0, 780.0, 495.0)),
  (MessageContent, Rect.fromLTRB(36.0, 461.0, 764.0, 491.0))]]
Actual: [[(MessageWithPossibleSender, Rect.fromLTRB(20.0, 187.0, 780.0, 259.0)),
(MessageContent, Rect.fromLTRB(36.0, 225.0, 764.0, 255.0))],
[(MessageWithPossibleSender, Rect.fromLTRB(20.0, 259.0, 780.0, 457.0)),
(MessageContent, Rect.fromLTRB(34.0, 263.0, 764.0, 430.0)),
(Star, Rect.fromLTRB(764.0, 270.3, 780.0, 286.3)),
(ReactionChipsList, Rect.fromLTRB(34.0, 430.0, 764.0, 453.0))],
[(MessageWithPossibleSender, Rect.fromLTRB(20.0, 457.0, 780.0, 495.0)),
(MessageContent, Rect.fromLTRB(36.0, 461.0, 764.0, 491.0))]]
Which: at [<1>][<1>] is <(MessageContent, Rect.fromLTRB(34.0, 263.0, 764.0, 430.0))>
which does not equal <(MessageContent, Rect.fromLTRB(36.0, 263.0, 764.0, 430.0))>
```

Much clearer! The developer can immediately see that the discrepancy is that the left edge of MessageContent is 2px too far to the left. (Which indeed it is, with the artificial error I induced as described above which makes the marker icon 2px too small.)

PIG208 replied:
Agreed with the change. When I was testing Flutter uglyprinted the maps. This made me I wonder if they can be displayed better. I remember vaguely that pytest highlights the differences between two dicts in a diff-like fashion, making it easier to spot the differences.

gnprice replied:
Yeah, I think some other nice improvements could be:
* better pretty-print the expected and actual data — above you can see both maps and lists lack indentation to show the structure, in particular
* fancier diff comparison at an algorithmic level — detecting things like "an element was inserted" in the middle of a list, for example
* printing actual vs. expected in diff-style formatting, with `-`/`+` lines and/or with red and green colors

The last one is a matter of preference which people might differ on. The other two are probably uncontroversial improvements… given a solid implementation of them. Both can be complex to get right — pretty-printing in particular can sound simple but turns out to be tricky when you try to really pin it down.

gnprice replied:
(And all three of those are found in some popular other test frameworks — pytest is a good comparison, and the Jest framework for JS is another. Pretty sure each of the three is found in one or both of those.)

PIG208 commented:
<details>
<summary>Latest screenshots</summary>

| Moved | Edited |
| - | - |
| ![image](https://github.com/user-attachments/assets/76a6b0ae-4b64-4555-b182-449fbaab78ae) | ![Screenshot from 2024-07-16 14-09-08](https://github.com/user-attachments/assets/d825b384-8e23-46d0-9ce7-f0b3716089de) |
| ![image](https://github.com/user-attachments/assets/83f75587-11c6-4f69-88d4-e57fec55265a) | ![Screenshot from 2024-07-16 14-11-43](https://github.com/user-attachments/assets/86833d96-d459-4f91-91fb-5498abaaee76) |
| ![Screenshot from 2024-07-16 14-15-49](https://github.com/user-attachments/assets/00987ee4-0a43-48cd-9ed3-944110e4eba2) | ![image](https://github.com/user-attachments/assets/4e1a22fb-8fab-4343-ace2-624225db5cb1) |
| ![image](https://github.com/user-attachments/assets/9c8dbe36-1653-44de-b39d-b8e25e80894f) | ![Screenshot from 2024-07-16 14-16-04](https://github.com/user-attachments/assets/f739e65c-cbdc-449d-82fb-8fb6946e4d33) |


</details>

PIG208 commented:
Both icons are 14x14. `message_moved` floats from the baseline while `edited` aligns its bottom to it. It differs from the design:

| Moved | Edited |
| - | - |
| ![Screenshot from 2024-07-16 14-24-52](https://github.com/user-attachments/assets/88086ea6-4f3f-4cff-8e1e-1c117e5e8f8c) | ![Screenshot from 2024-07-16 14-24-38](https://github.com/user-attachments/assets/4e19f6c5-d3a8-482a-857a-7586f808eebb) |

I'm going to apply an offset to shift them from the baseline to the desired position.

PIG208 commented:
<details>
<summary>Latest screenshots (with the offset)</summary>

| Moved | Edited |
| - | - |
| ![Screenshot from 2024-07-16 14-31-57](https://github.com/user-attachments/assets/cb6de750-d352-49d7-9202-33efd7167073) | ![Screenshot from 2024-07-16 14-31-42](https://github.com/user-attachments/assets/63e6c533-add9-4424-8cce-378d1b877e63) |
| ![Screenshot from 2024-07-16 14-32-20](https://github.com/user-attachments/assets/b8c0d2c6-35df-4d50-89d9-01f1739baf2e) | ![Screenshot from 2024-07-16 14-32-25](https://github.com/user-attachments/assets/2f18553b-182f-41d6-93bd-3a8dfa745819) |
| ![Screenshot from 2024-07-16 14-34-04](https://github.com/user-attachments/assets/5f22ab5b-b730-4097-bfd5-97ffeffadbeb) | ![Screenshot from 2024-07-16 14-34-41](https://github.com/user-attachments/assets/3b7aa12c-5b8e-4cf7-b1cc-1b2cb9136a12) |
| ![Screenshot from 2024-07-16 14-35-32](https://github.com/user-attachments/assets/147e17c3-37ed-4b3a-8a08-d5deefcfa0be) | ![Screenshot from 2024-07-16 14-35-11](https://github.com/user-attachments/assets/d4b65a74-ab2d-437a-9e7f-2152aeb1977d) |




</details>

PIG208 commented:
<details>
<summary>
Latest screenshots (with the 16x16 icons):
</summary>

| Moved | Edited |
| - | - |
| ![Screenshot from 2024-07-16 19-34-12](https://github.com/user-attachments/assets/d0315aaa-7622-489d-991e-a615536d05d2) | ![Screenshot from 2024-07-16 19-34-07](https://github.com/user-attachments/assets/f699e60e-5b07-4778-9b09-0a5caae7e344) |
| ![Screenshot from 2024-07-16 19-34-44](https://github.com/user-attachments/assets/7c82d22e-b7c9-4c60-8a5f-1d2efb2065e8) | ![Screenshot from 2024-07-16 19-34-30](https://github.com/user-attachments/assets/6845761a-d259-4c26-8869-b17b9515b922) |
| ![Screenshot from 2024-07-16 19-35-54](https://github.com/user-attachments/assets/415a5ba3-8e6b-4f5a-b992-5fb5669c8bda) | ![Screenshot from 2024-07-16 19-35-46](https://github.com/user-attachments/assets/c9b11e5d-ec10-4500-8bd1-979677158205) |

</details>

PIG208 commented:
With the new icons, not sure if the offsets are needed anymore.
This should be ready for review.

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -964,25 +963,21 @@ class MessageWithPossibleSender extends StatelessWidget {
-            const SizedBox(width: 16),
-            Expanded(
-              child: Column(
+          EditStateMarker(
+            message: message,
+            children: [
```
bump :slightly_smiling_face:

PIG208 replied:
Oops. Fixed!

gnprice commented:
Hmm, here's a screenshot of what I see:
![image](https://github.com/user-attachments/assets/d9b0f975-5623-4ca1-9753-00fe34c13baa)

So visually the bottom of the icon is higher than the baseline of the text, and the top of the icon is much higher than the ascender line of the text.

That's on my usual display settings, which have the font size set to small (as well as the "display size", which effectively just means my screen has more logical pixels). On default settings, I see:
![image](https://github.com/user-attachments/assets/0ed524ff-d50c-4e68-af9e-9567e980d59a)

So the same thing, but a little less so, because the icons aren't as much taller than the text as they are at my usual settings.

gnprice commented:
---

On our screen-sharing call today, when you were running the app locally on Linux, it looked like you were seeing the symptom of #735. That suggests that your system has the text scaling set to >1x, i.e. the fonts are scaled up. If I crank the text size on my phone a couple of notches above default, I reproduce a layout more like your screenshots above.

When making screenshots for a piece of layout that involves text, it'd be good to do so with a 1x text scale. It can be helpful to try smaller or larger scaling too, but 1x should be the baseline.

I'm not sure what platform setting might be governing the text scaling Flutter picks up from the platform on Linux. If you can't readily find that setting, you can always add a hack locally in `lib/app.dart` that forces the scaling to 1x — for example a `MediaQuery` around the `ZulipApp`.

gnprice commented:
For the vertical alignment for these icons, there are probably fancy things we can do to get it just right for a range of situations — but let's skip that for the present, while we move on to other issues from our tracker.

For now, let's just try to get it so that at 1x text scaling, when the message starts with a plain paragraph, the top of the edit icon doesn't stick out above the ascender line of the text, or anyway not by much more than the bottom of the icon sticks out below the baseline of the text. If we do that by a fixed few px of vertical adjustment, that'll be fine.

gnprice commented:
Other than that layout issue, the code here all looks good modulo this small comment above: https://github.com/zulip/zulip-flutter/pull/762#discussion_r1680182894

Thanks for the revision!

PIG208 commented:
Updated the PR to add the vertical adjustments and replace the outdated icon font. Thanks for noticing that scaling issue!

<details>

<summary>With the offset, default text scaling (fd38fb86f418b8243df77f900a3e4df4b090660d)</summary>

| Moved | Edited |
| - | - |
| ![Screenshot from 2024-07-16 20-23-35](https://github.com/user-attachments/assets/fbde448f-c71a-4e2e-8be9-3160a031b7c3) | ![Screenshot from 2024-07-16 20-23-27](https://github.com/user-attachments/assets/4b380119-adfc-47b4-8e48-c8b5f093be84) |
| ![Screenshot from 2024-07-16 20-23-42](https://github.com/user-attachments/assets/ebec6d0f-82cd-4ae7-adf8-e45ba68daa5e) | ![Screenshot from 2024-07-16 20-23-47](https://github.com/user-attachments/assets/656b17e8-3b65-4b77-8d19-36fd7b0226cd) |
| ![Screenshot from 2024-07-16 20-24-05](https://github.com/user-attachments/assets/9c76ca87-7aeb-4cac-a6d6-017b53e9bdcd) | ![Screenshot from 2024-07-16 20-24-00](https://github.com/user-attachments/assets/f7a8a2c5-400b-4694-8bbd-2e8dbbe41d4b) |

</details>

gnprice commented:
Great, thanks! Looks good — merging.

I added a comment about those vertical offsets:
```diff
+        // These offsets are chosen ad hoc, but give a good vertical alignment
+        // of the icons with the first line of the message, when the message
+        // begins with a paragraph, at default text scaling.  See:
+        //   https://github.com/zulip/zulip-flutter/pull/762#issuecomment-2232041922
         offset = const Offset(0, 2);
```

Also edited the PR description so that #171 doesn't get accidentally closed.
